### PR TITLE
Remove PPromise usage from NSFW IPC

### PR DESCRIPTION
**Pull request by joaomoreno**
Related to #53487

@bpasero This removes PPromise usage from the NSFWWatcherService. Give it a good test drive, especially wrt the process crashing and restarting everything again. Everything should work as before.

If good, I'll make exactly the same adoption for Chokidar, since both worlds seem identical.

### CODE REVIEW

File: src/vs/workbench/services/files/node/watcher/nsfw/watcherService.ts [hunk 1]
```diff
@@ -16,15 +15,16 @@ import { IWorkspaceContextService } from 'vs/platform/workspace/common/workspace
 import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
 import { IDisposable, dispose } from 'vs/base/common/lifecycle';
 import { Schemas } from 'vs/base/common/network';
-import { isPromiseCanceledError } from 'vs/base/common/errors';
+import { filterEvent } from 'vs/base/common/event';
+import { isError } from 'util';
```
Comment (bpasero) [inline src/vs/workbench/services/files/node/watcher/nsfw/watcherService.ts]: @joaomoreno did you mean to use `errors.isError` here?

**Review by bpasero — CHANGES_REQUESTED** [review]
@joaomoreno left some comments but could not test after all due to this error:

```
  ERR event is not a function: TypeError: event is not a function
    at file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/event.js:441:20
    at file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/event.js:441:20
    at FileWatcher.startWatching (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/workbench/services/files/node/watcher/nsfw/watcherService.js:51:13)
    at RemoteFileService.FileService.setupFileWatching (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/workbench/services/files/electron-browser/fileService.js:149:99)
    at file:///Users/bpasero/Development/Microsoft/monaco/out/vs/workbench/services/files/electron-browser/fileService.js:90:23
    at Object.notifySuccess [as _notify] (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1191:59)
    at Object.enter (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:867:30)
    at Promise_ctor._run (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1089:29)
    at Promise_ctor._completed (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1057:18)
    at Barrier.open (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/async.js:339:18)
    at LifecycleService.set [as phase] (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/platform/lifecycle/electron-browser/lifecycleService.js:95:54)
    at file:///Users/bpasero/Development/Microsoft/monaco/out/vs/workbench/electron-browser/shell.js:70:50
    at _Base.Class.derive._creator.CompletePromise_done [as done] (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1545:34)
    at Object.notifySuccess [as _notify] (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1201:48)
    at Object.enter (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:867:30)
    at Promise_ctor._run (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1089:29)
    at Promise_ctor._completed (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1057:18)
    at argDone (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1834:37)
    at Promise.then.errors.(anonymous function) (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1859:78)
    at Object.notifySuccess [as _notify] (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1191:59)
    at Object.enter (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:867:30)
    at _Base.Class.derive._creator._run (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1089:29)
    at _Base.Class.derive._creator._completed (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1057:18)
    at _Base.Class.derive._creator.CompletePromise_then [as then] (file:///Users/bpasero/Development/Microsoft/monaco/out/vs/base/common/winjs.base.js:1587:49)
```

File: src/vs/base/common/errors.ts [hunk 2]
```diff
@@ -242,6 +242,10 @@ export interface IErrorWithActions {
 	actions?: IAction[];
 }
 
+export function isError(obj: any): obj is Error {
```
[the remark this answers is not in the document]
Comment (joaomoreno) [reply src/vs/base/common/errors.ts]: I was referencing the wrong one...

File: src/vs/workbench/services/files/node/watcher/nsfw/nsfwWatcherService.ts [hunk 3]
```diff
@@ -35,20 +36,15 @@ export class NsfwWatcherService implements IWatcherService {
...
-		this._watcherPromise = new TPromise<void>((c, e, p) => {
-			this._errorCallback = e;
-			this._progressCallback = p;
+	private _onWatchEvent = new Emitter<watcher.IRawFileChange[] | Error>();
+	readonly onWatchEvent = this._onWatchEvent.event;
 
-		});
-		return this._watcherPromise;
+	watch(options: IWatcherOptions): Event<watcher.IRawFileChange[] | Error> {
+		this._verboseLogging = true;
```
[the remark this answers is not in the document]
Comment (joaomoreno) [reply src/vs/workbench/services/files/node/watcher/nsfw/nsfwWatcherService.ts]: Yup, thought you did that on purpose.

File: src/vs/workbench/services/files/node/watcher/nsfw/watcherService.ts (hunk 1 above)
Comment (joaomoreno) [reply src/vs/workbench/services/files/node/watcher/nsfw/watcherService.ts]: YEAH!

**Comment by joaomoreno** [comment]
Pushed review comment changes.

**Review by nzec — APPROVED** [review]

**Review by bpasero — CHANGES_REQUESTED** [review]
@joaomoreno I tested basic things like file changes and simulating a crash of the watcher and it works fine. however this does not work: if I manually trigger the ENOSPC error after a short timeout from [here](https://github.com/Microsoft/vscode/blob/0679d90bca576aaa6398e26e10dd06f6f64eeaa5/src/vs/workbench/services/files/node/watcher/nsfw/nsfwWatcherService.ts#L73) I do not get a notification. Compared to stable:

image

**Comment by joaomoreno** [comment]
Great catch. Pushed fix!

**Review by bpasero — APPROVED** [review]
LGTM, thanks!

**Pull request merged**